docs/src/modules/ROOT/pages/constraints-and-score/performance.adoc

```diff
@@ -179,35 +222,76 @@ Let's instead remove them from the cross-product entirely.
 For the first lesson in the join, also called “left,”
 we put the cancellation check before the join like so:
 
-[source, java]
+[tabs]
+====
+Java::
++
+[source, java, options="nowrap"]
 ----
     constraintFactory.forEach(Lesson.class)
         .filter(lesson -> !lesson.isCancelled())
         .join(Lesson.class)
         .filter((leftLesson, rightLesson) ->
             !rightLesson.isCancelled()
-            && leftLesson.getTeacher() == rightLesson.getTeacher()
+            && leftLesson.getTeacher().equals(rightLesson.getTeacher())
```

Comment:
Technically, I think the previous example was correct too; the model should ensure there are never two instances which equal. It's a problem fact, those don't need identity.

Reply:
But it was inconsistent; it used `equals` in one place, and `==` in another.